### DISCUSSION TITLE
pylint unable to verify fastapi import

### DISCUSSION THREAD

**Discussion by gamesbook** [fastapi/fastapi#6621] [Questions]
Using:

FastAPI 0.63.0
Python 3.7.7
OS: Mint 19

I need to prevent pylint from reporting E0401 error. Below are the steps taken to reproduce the problem in an isolated virtual environment (command + output is shown).

```
derek@pc:~/tmp$ mkvirtualenv slow
created virtual environment CPython3.7.7.final.0-64 in 182ms
  creator CPython3Posix(dest=/home/derek/.envs/slow, clear=False, global=False)
  seeder FromAppData(download=False, pip=latest, setuptools=latest, wheel=latest, via=copy, app_data_dir=/home/derek/.local/share/virtualenv/seed-app-data/v1.0.1)
  activators BashActivator,CShellActivator,FishActivator,PowerShellActivator,PythonActivator,XonshActivator
virtualenvwrapper.user_scripts creating /home/derek/.envs/slow/bin/predeactivate
virtualenvwrapper.user_scripts creating /home/derek/.envs/slow/bin/postdeactivate
virtualenvwrapper.user_scripts creating /home/derek/.envs/slow/bin/preactivate
virtualenvwrapper.user_scripts creating /home/derek/.envs/slow/bin/postactivate
virtualenvwrapper.user_scripts creating /home/derek/.envs/slow/bin/get_env_details
(slow) derek@pc:~/tmp$ pip freeze
(slow) derek@pc:~/tmp$ pip install fastapi
Collecting fastapi
  Using cached fastapi-0.63.0-py3-none-any.whl (50 kB)
Collecting starlette==0.13.6
  Using cached starlette-0.13.6-py3-none-any.whl (59 kB)
Collecting pydantic<2.0.0,>=1.0.0
  Using cached pydantic-1.8.1-cp37-cp37m-manylinux2014_x86_64.whl (10.1 MB)
Collecting typing-extensions>=3.7.4.3
  Downloading typing_extensions-3.10.0.0-py3-none-any.whl (26 kB)
Installing collected packages: starlette, typing-extensions, pydantic, fastapi
Successfully installed fastapi-0.63.0 pydantic-1.8.1 starlette-0.13.6 typing-extensions-3.10.0.0
(slow) derek@pc:~/tmp$ nano test.py
```
The test.py file simply has one line:
```
import fastapi
```
Now run pylint on the file:
```
(slow) derek@pc:~/tmp$ pylint test.py
************* Module test
test.py:1:0: C0114: Missing module docstring (missing-module-docstring)
test.py:1:0: E0401: Unable to import 'fastapi' (import-error)
test.py:1:0: W0611: Unused import fastapi (unused-import)
```
If I try to install and import another library, I do not have this error:
```
(slow) derek@pc:~/tmp$ pip install requests
Collecting requests
  Using cached requests-2.25.1-py2.py3-none-any.whl (61 kB)
Collecting idna<3,>=2.5
  Using cached idna-2.10-py2.py3-none-any.whl (58 kB)
Collecting urllib3<1.27,>=1.21.1
  Using cached urllib3-1.26.4-py2.py3-none-any.whl (153 kB)
Collecting certifi>=2017.4.17
  Using cached certifi-2020.12.5-py2.py3-none-any.whl (147 kB)
Collecting chardet<5,>=3.0.2
  Using cached chardet-4.0.0-py2.py3-none-any.whl (178 kB)
Installing collected packages: idna, urllib3, certifi, chardet, requests
Successfully installed certifi-2020.12.5 chardet-4.0.0 idna-2.10 requests-2.25.1 urllib3-1.26.4
```
Add the `import requests` to the test file.
```
(slow) derek@pc:~/tmp$ pylint test.py
************* Module test
test.py:1:0: C0114: Missing module docstring (missing-module-docstring)
test.py:1:0: E0401: Unable to import 'fastapi' (import-error)
test.py:1:0: W0611: Unused import fastapi (unused-import)
test.py:2:0: W0611: Unused import requests (unused-import)
```

**Comment by me-v2** [comment]
indeed!   
`/Users/spacex/PycharmProjects/fastatpi-start/main.py:1: [E0401(import-error), ] Unable to import 'fastapi'`

**Comment by waynerv** [comment]
Can't reproduce this. I installed pylint in the same virtual environment.
```
❯ pylint test.py
************* Module test
test.py:1:0: C0114: Missing module docstring (missing-module-docstring)
test.py:1:0: W0611: Unused import fastapi (unused-import)
```

**Comment by gamesbook** [comment]
What is the underlying reason why an error would be triggered for one library and not another? How is fastapi different from, for example, requests?  

Alternatively, what setup is needed to ensure that this error is not raised?

**Comment by chbndrhnns** [comment]
Is pylint installed in that venv, as well? In my experience this makes a difference.

**Comment by gamesbook** [comment]
@chbndrhnns Yes, you are correct.  It works when you install local pylint.

Unfortunately, I am still left with the issue that caused me to go down this path in the first place - the fact that Spyder (Python IDE) does not seem able to make use of pylint when its installed this way; see:

https://github.com/spyder-ide/spyder/issues/2378

and 

https://stackoverflow.com/questions/63932897/pylint-in-spyder-is-raising-e401-unable-to-import-whithin-conda-environment

And especially the last paragraph from the above:

_Spyder always uses pylint from my dedicated slim spyder environment (conda) and hence does not find the modules in the actual development environment. Running pylint from terminal in the actual environment works, but is not a nice work-around_

**Comment by chbndrhnns** [comment]
> Running pylint from terminal in the actual environment works, but is not a nice work-around

This workaround is even required for PyCharm in cases like pre-commit hooks. :/

**Comment by gamesbook** [comment]
It does appear that VS-Code will use pylint from the local (virtual)
environment rather than global - but I do not have that running to confirm.

**Comment by tiangolo** [comment]
Thanks for the help here everyone! 👏  🙇 

Thanks for reporting back and closing the issue @gamesbook  👍

---

For completeness, if Spyder uses its own version of Pylint instead of the one you install in the local environment for your project, then it would actually be Spyder's fault/error. That would probably be a report/request there.

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
